### PR TITLE
Add test cases for taintbasedevictions

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**What this PR does / why we need it**:
Add unit tests for basic taint based evictions.
/cc @Huang-Wei @k82cn @bsalamat 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by k82cn** [comment]
Thanks, I'm going to take time to review this :)

/assign

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 1]
```diff
@@ -1977,6 +1976,122 @@ func TestMonitorNodeHealthMarkPodsNotReady(t *testing.T) {
 	}
 }
 
+// TestEvictPodsByTaints, ensures we just have a NoExecute taint applied to node.
+// NodeController is just responsible for enqueuing the node to tainting queue from which taint manager picks up
+// and evicts the pods on the node.
+func TestEvictPodsByTaints(t *testing.T) {
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: change func name to "TestApplyNoExecuteTaints"?

**Review by Huang-Wei — COMMENTED** [review]
Could you add more cases to cover the scenario that by giving nodes having NoExecute taints, once we got a Ready=true condition, those NoExecute taints are expected to take off those nodes?

Otherwise lgtm. Thanks @ravisantoshgudimetla !

**Review by bsalamat — COMMENTED** [review]
LGTM, but I would like someone more familiar with the logic to take a look.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 2]
```diff
@@ -1977,6 +1976,157 @@ func TestMonitorNodeHealthMarkPodsNotReady(t *testing.T) {
...
+	node3, err := fakeNodeHandler.Get("node3", metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("Can't get current node3...")
+		return
+	}
+	if !taintutils.TaintExists(node3.Spec.Taints, NotReadyTaintTemplate) {
+		t.Errorf("Can't find taint %v in %v", NotReadyTaintTemplate, node3.Spec.Taints)
+	}
+
+	// Make node3 healthy again.
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: can we also make `node0` back to healthy, and check if unreachable taint is taken off also?

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 3]
```diff
@@ -1977,6 +1976,157 @@ func TestMonitorNodeHealthMarkPodsNotReady(t *testing.T) {
...
+							LastHeartbeatTime:  metav1.Date(2017, 1, 1, 12, 0, 0, 0, time.UTC),
+							LastTransitionTime: metav1.Date(2017, 1, 1, 12, 0, 0, 0, time.UTC),
+						},
+					},
+				},
+			},
+			// NotReady Taint with NoExecute effect should be applied to this node.
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node3",
```
Comment (Huang-Wei) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: where is `node2`...

Comment (ravisantoshgudimetla) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: node3 preempted node2, lol, I missed it. I will change it.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 2 above)
Comment (ravisantoshgudimetla) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: I think, we are exercising the same codepath in both the cases(node0 and node3), so no need to check explicitly.

**Review by Huang-Wei — APPROVED** [review]
/lgtm

**Comment by spiffxp** [comment]
/milestone v1.13
I'm adding this to the v1.13 milestone since it relates to https://github.com/kubernetes/features/issues/166 which is currently planned for v1.13

(also, thanks for adding unit test coverage!)

**Comment by Huang-Wei** [comment]
@ravisantoshgudimetla I think #54297 can be marked as a duplicate of this?

**Pull request merged**